rcl/src/rcl/logging_rosout.c

```diff
@@ -292,3 +355,133 @@ void rcl_logging_rosout_output_handler(
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(full_sublogger_name, RCL_RET_INVALID_ARGUMENT);
+
+  if (logger_name[0] == '\0' || sublogger_name[0] == '\0') {
+    RCL_SET_ERROR_MSG("logger name or sub-logger name can't be empty.");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  *full_sublogger_name = rcutils_format_string(
+    __rosout_allocator, "%s%s%s",
+    logger_name, RCUTILS_LOGGING_SEPARATOR_STRING, sublogger_name);
+  if (NULL == *full_sublogger_name) {
+    RCL_SET_ERROR_MSG("Failed to allocate a full sublogger name.");
+    return RCL_RET_BAD_ALLOC;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_logging_rosout_add_sublogger(
+  const char * logger_name, const char * sublogger_name)
+{
+  if (!__is_initialized) {
+    return RCL_RET_OK;
+  }
+  rcl_ret_t status = RCL_RET_OK;
+  char * full_sublogger_name = NULL;
+  rosout_map_entry_t entry;
+
+  status =
+    _rcl_logging_rosout_get_sublogger_name(logger_name, sublogger_name, &full_sublogger_name);
+  if (RCL_RET_OK != status) {
+    // Error already set
+    return status;
+  }
+
+  rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("The entry of logger '%s' not exist.", logger_name);
+    status = RCL_RET_ERROR;
+    goto cleanup;
+  }
+
+  if (rcutils_hash_map_key_exists(&__logger_map, &full_sublogger_name)) {
+    status = RCL_RET_SUBLOGGER_ALREADY_EXIST;
+    goto cleanup;
+  }
+
+  status = rcl_ret_from_rcutils_ret(
+    rcutils_hash_map_set(&__logger_map, &full_sublogger_name, &entry));
+  if (RCL_RET_OK != status) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Failed to add publisher to map for logger '%s'.", full_sublogger_name);
+    goto cleanup;
+  }
+
+  status = rcl_ret_from_rcutils_ret(
+    rcutils_array_list_add(&__sublogger_names, &full_sublogger_name));
+  if (RCL_RET_OK != status) {
+    goto cleanup;
+  }
+
+  return status;
+
+cleanup:
+  __rosout_allocator.deallocate(full_sublogger_name, __rosout_allocator.state);
+  return status;
+}
+
+rcl_ret_t
+rcl_logging_rosout_remove_sublogger(
+  const char * logger_name, const char * sublogger_name)
+{
+  if (!__is_initialized) {
+    return RCL_RET_OK;
+  }
+  rcl_ret_t status = RCL_RET_OK;
+  char * full_sublogger_name = NULL;
+  size_t array_size;
+  char * sublogger_name_removed = NULL;
+
+  status =
+    _rcl_logging_rosout_get_sublogger_name(logger_name, sublogger_name, &full_sublogger_name);
+  if (RCL_RET_OK != status) {
+    // Error already set
+    return status;
+  }
+
+  // remove the entry from the map
+  if (!rcutils_hash_map_key_exists(&__logger_map, &full_sublogger_name)) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("Sub-logger '%s' not exist.", full_sublogger_name);
+    __rosout_allocator.deallocate(full_sublogger_name, __rosout_allocator.state);
```

Comment:
Like `rcl_logging_rosout_add_sublogger`, we should have a `cleanup` label and go there when we fail.

This will also allow us to unravel the cleanup below, so we can `goto cleanup` during failures.